inc/marketplace/controller.class.php

```diff
@@ -0,0 +1,175 @@
+<?php
+/**
+ * ---------------------------------------------------------------------
+ * GLPI - Gestionnaire Libre de Parc Informatique
+ * Copyright (C) 2015-2018 Teclib' and contributors.
+ *
+ * http://glpi-project.org
+ *
+ * based on GLPI - Gestionnaire Libre de Parc Informatique
+ * Copyright (C) 2003-2014 by the INDEPNET Development Team.
+ *
+ * ---------------------------------------------------------------------
+ *
+ * LICENSE
+ *
+ * This file is part of GLPI.
+ *
+ * GLPI is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * GLPI is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with GLPI. If not, see <http://www.gnu.org/licenses/>.
+ * ---------------------------------------------------------------------
+ */
+
+namespace Glpi\Marketplace;
+
+if (!defined('GLPI_ROOT')) {
+   die("Sorry. You can't access directly to this file");
+}
+
+
+use Glpi\Marketplace\Api\Plugins as PluginsApi;
+use \wapmorgan\UnifiedArchive\UnifiedArchive;
+
+class Controller extends \CommonGLPI {
+   protected $plugin_key = "";
+
+   static $rightname = 'config';
+   static $api       = null;
+
+
+   function __construct(string $plugin_key = "") {
+      $this->plugin_key = $plugin_key;
+   }
+
+   static function getAPI(): PluginsApi {
+      return self::$api ?? (self::$api = new PluginsApi());
+   }
+
+
+   function downloadPlugin():int {
+      $api      = self::getAPI();
+      $plugin   = $api->getPlugin($this->plugin_key, true);
+      $url      = $plugin['installation_url'] ?? "";
+      $file     = file_get_contents($url);
```

Comment:
This may use a huge amount of memory depending on archive size.
Using cURL to save it directly inside a file would consume less memory (code snippets here: https://stackoverflow.com/questions/1006604/saving-file-using-curl-and-php).